File: src/main/java/ch/epfl/dias/calcite/adapter/pelago/reporting/TimeKeeper.java
package ch.epfl.dias.calcite.adapter.pelago.reporting;

import ch.epfl.dias.repl.Repl;
import java.sql.Timestamp;

/** Global singleton for time measurements */
public class TimeKeeper {
    public static final TimeKeeper INSTANCE = new TimeKeeper();

    private long tPlanToJson;
    private long tPlanning;
    private long tExecutor;
    private long tCodegen;
    private long tDataLoad;
    private long tCodeOpt;
    private long tCodeOptAndLoad;
    private long tExec;

    private Timestamp lastTimestamp;

    private TimeKeeper() {}

    public void addTexec(long time_ms) {
        tExec = time_ms;
    }

    public void addTcodegen(long time_ms) {
        tCodegen = time_ms;
    }

    public void addTdataload(long time_ms) {
        tDataLoad = time_ms;
    }

    public void addTcodeopt(long time_ms) {
        tCodeOpt = time_ms;
    }

    public void addTcodeoptnload(long time_ms) {
        tCodeOptAndLoad = time_ms;
    }

    public void addTplan2json(long time_ms) {
        tPlanToJson = time_ms;
    }

    public void addTexecutorTime(PelagoTimeInterval interval) {
        tExecutor = interval.getDifferenceMilli();
    }

    public void addTplanning(PelagoTimeInterval interval) {
        tPlanning = interval.getDifferenceMilli();
    }

    public void addTimestamp(){
        lastTimestamp = new Timestamp(System.currentTimeMillis());
    }

    public void refreshTable() {
        TimeKeeperTable.addTimings(tExecutor + tPlanning + tPlanToJson, tPlanning, tPlanToJson, tExecutor, tCodegen, tDataLoad, tCodeOpt, tCodeOptAndLoad, tExec, lastTimestamp);
    }


    @Override
    public String toString() {
        String format;
        if (Repl.timingscsv()) {
            format = "Timings,%d,%d,%d,%d,%d,%d,%d,%d,%d";
        } else {
            format = "Total time: %dms, "
                    + "Planning time: %dms, "
                    + "Flush plan time: %dms, "
                    + "Total executor time: %dms, "
                    + "Codegen time: %dms, "
                    + "Data Load time: %dms, "
                    + "Code opt time: %dms, "
                    + "Code opt'n'load time: %dms, "
                    + "Execution time: %dms";
        }
        return String.format(format, tExecutor + tPlanning + tPlanToJson, tPlanning, tPlanToJson, tExecutor,
                                tCodegen, tDataLoad, tCodeOpt, tCodeOptAndLoad, tExec, lastTimestamp);
    }
}

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/reporting/TimeKeeper.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package ch.epfl.dias.calcite.adapter.pelago.reporting

import ch.epfl.dias.repl.Repl
import java.sql.Timestamp

/** Global singleton for time measurements */
object TimeKeeper { var INSTANCE = new TimeKeeper }
class TimeKeeper private () {
  private var tPlanToJson: java.lang.Long = 0L
  private var tPlanning: java.lang.Long = 0L
  private var tExecutor: java.lang.Long = 0L
  private var tCodegen: java.lang.Long = 0L
  private var tDataLoad: java.lang.Long = 0L
  private var tCodeOpt: java.lang.Long = 0L
  private var tCodeOptAndLoad: java.lang.Long = 0L
  private var tExec: java.lang.Long = 0L
  private var lastTimestamp: Timestamp = null

  private def reset(): Unit = TimeKeeper.INSTANCE = new TimeKeeper
  def addTexec(time_ms: Long): Unit = tExec = time_ms
  def addTcodegen(time_ms: Long): Unit = tCodegen = time_ms
  def addTdataload(time_ms: Long): Unit = tDataLoad = time_ms
  def addTcodeopt(time_ms: Long): Unit = tCodeOpt = time_ms
  def addTcodeoptnload(time_ms: Long): Unit = tCodeOptAndLoad = time_ms
  def addTplan2json(time_ms: Long): Unit = tPlanToJson = time_ms
  def addTexecutorTime(interval: PelagoTimeInterval): Unit =
    tExecutor = interval.getDifferenceMilli
  def addTplanning(interval: PelagoTimeInterval): Unit =
    tPlanning = interval.getDifferenceMilli
  def addTimestamp(): Unit =
    lastTimestamp = new Timestamp(System.currentTimeMillis)
  def refreshTable(
      query: String,
      cmd_type: String,
      hwmode: String,
      plan: String,
      query_name: String
  ): Unit = {
    if (lastTimestamp == null) addTimestamp()
    TimeKeeperTable.addTimings(
      tExecutor + tPlanning + tPlanToJson,
      tPlanning,
      tPlanToJson,
      tExecutor,
      tCodegen,
      tDataLoad,
      tCodeOpt,
      tCodeOptAndLoad,
      tExec,
      lastTimestamp,
      query,
      cmd_type,
      hwmode,
      plan,
      query_name
    )
    reset()
  }

  override def toString: String = {
    val format: java.lang.String = {
      if (Repl.timingscsv) "Timings,%d,%d,%d,%d,%d,%d,%d,%d,%d,%s,%s"
      else
        "Total time: %dms, " + "Planning time: %dms, " + "Flush plan time: %dms, " + "Total executor time: %dms, " + "Codegen time: %dms, " + "Data Load time: %dms, " + "Code opt time: %dms, " + "Code opt'n'load time: %dms, " + "Execution time: %dms"
    }
    String.format(
      format,
      (tExecutor + tPlanning + tPlanToJson).asInstanceOf[java.lang.Long],
      tPlanning,
      tPlanToJson,
      tExecutor,
      tCodegen,
      tDataLoad,
      tCodeOpt,
      tCodeOptAndLoad,
      tExec,
      lastTimestamp
    )
  }
}
